# WGS analysis pipeline
WGS analysis pipeline. Can handle both WGS and WES data.

The whole pipeline use singularity images and will pull images from singularity libraries when needed

## How to run
The pipeline can be run directly using Nextflow >= v20.10.

```
nextflow WGS_analysis.nf -profile cluster --operation align --input input_file.txt --mode WGS --ped ped_file.ped --cohort_id cohort_name --outdir results
```

The pipeline automatically infer the number of samples in the cohort from your input file. When more than one sample is present, small variants and structural variants from all samples are merged in cohort wide VCF files. 

Eventually update `singularity_cachedir` variable in `nextflow.config` to point to a proper folder where singularity images are stored / will be downloaded

A bash script `Run_pipeline.sh` is provided for convenience to set environmental variables for singularity cache and temp directories. The `singularity_dir` variable in the script must match `singularity_cachedir` variable in `nextflow.config`.

### Arguments

```
operation   :   align or call_variants
mode        :   WGS or WES
input       :   tab-separated file describing input files. 
                The exact format depends on operation requested (see below)
ped         :   standard PED file containing all samples
cohort_id   :   a arbitrary name for the cohort files generated
outdir      :   output folder for results
```

## Input files format
### PED file
A standard tab-separated PED file without header, describing all samples provided in the input file. All sample IDs must match between ped and input file. All samples must have sex defined.

```
family_ID   individual_ID   father_ID   mother_ID   sex(1=M,2=F)    status(1=unaff,2=aff,0=unknown)
```

### input file

Note that all files need to be specified using **absolute paths**

#### Operation: align
A 3 columns tab-separated file without header

```
sampleID1   s1_lane1_R1.fastq.gz    s1_lane1_R2.fastq.gz
sampleID1   s1_lane2_R1.fastq.gz    s1_lane2_R2.fastq.gz
sampleID2   s2_lane2_R1.fastq.gz    s2_lane2_R2.fastq.gz
```

Note that if a sample has been sequenced with multiple pairs of fastq files you need to add multiple lines for each pair of fastq files using the same sampleID. The pipeline will take care of the merge.

#### Operation: call_variants
A 5 columns tab-separated file without header.
This file is automatically generated when using `--operation align` (bam_files.txt) 

```
sampleID1   main_bam.bam    disc.bam    split.bam
sampleID2   main_bam.bam    disc.bam    split.bam
sampleID3   main_bam.bam    disc.bam    split.bam
```

## Pipeline components
1. Alignement and duplicate marking
    - BWA + samblaster + samtools
2. QC and coverage from BAM files
    - fastqc: reads stats
    - mosdepth: coverage
    - samtools flagstat / mapstat: alignment stats
    - somalier: ancestry, relatedness, sex check reports 
    - multiqc: interactive report
3. small variants
    - deepvariant: single sample calls
    - glnexus: gvcf merge 
4. structural variants
    - lumpy: structural variants events
    - CNVnator: CNV estimation
    - svtools: combine, merge and classify
5. repeat expansion detection
    - expansion hunter
6. ROH regions
    - bcftools ROH